### ISSUE TITLE
Hits are empty after 260 records

### ISSUE THREAD

**Issue by v-byte-cpu** [typesense/typesense#186]
## Description

After 260 returned records search **hits** array is empty.

## Steps to reproduce

Search query: 

```
/collections/videos/documents/search?q=man&query_by=title,categories,tags&per_page=20&page=14
```

Response:

```
{
  "facet_counts": [],
  "found": 10291,
  "hits": [],
  "out_of": 27000,
  "page": 14,
  "request_params": {
    "per_page": 20,
    "q": "man"
  },
  "search_time_ms": 8
}
```

Obviously, 10291 > 280 records.

The search query is performed on videos collection with the following schema:

```
{
    "name": "videos",
    "fields": [
        {
            "name": "title",
            "type": "string"
        },
        {
            "name": "previewUrl",
            "type": "string",
            "facet": true
        },
        {
            "name": "fullUrl",
            "type": "string",
            "facet": true
        },
        {
            "name": "views",
            "type": "int64"
        },
        {
            "name": "categories",
            "type": "string[]"
        },
        {
            "name": "tags",
            "type": "string[]"
        },
        {
            "name": "uploadDate",
            "type": "int64"
        },
        {
            "name": "duration",
            "type": "int32"
        }
    ],
    "default_sorting_field": "views"
}
```
## Expected Behavior

**Hits** field is not empty array

## Actual Behavior

**Hits** field is empty array

## Metadata

**Typsense Version**: 

v0.18.0

**OS**:

Not relevant, docker container

**Comment by Aaron23145** [comment]
I ran into this problem too when using Typesense in Ubuntu 20.04 LTS, with version 0.18.0 of Typesense.
From result 251, hits are always received empty.

Search query:
`/collections/products/documents/search?q=acid&query_by=name,description&page=26`

Response:
```
{
  "facet_counts": [],
  "found": 157259,
  "hits": [],
  "out_of": 1039110,
  "page": 26,
  "request_params": {
    "per_page": 10,
    "q": "acid"
  },
  "search_time_ms": 42
}
```

Schema:
```
{
  'name': 'products',
  'fields': [
    {'name': 'internal_id', 'type': 'int64'},
    {'name': 'internal_ref', 'type': 'string'},
    {'name': 'name', 'type': 'string'},
    {'name': 'description', 'type': 'string'},
    [... More irrelevant fields]
  ],
  'default_sorting_field': 'internal_id'
}
```

However, I also noticed that if I only query by one field, everything seems to work perfectly.

Search query (this one works):
`/collections/products/documents/search?q=acid&query_by=name&page=26`

With the last serach query I'm receiving what I'm expecting, a response with all the hits.

**Comment by jasonbosco** [comment]
@Aaron23145 Good observation about this only affecting multi-field searches! 

@v-byte-cpu @Aaron23145 We introduced a performance optimization in `v0.18.0` for multi-field searches that uses the `max_hits` search parameter to narrow down the search space. This value is set to 250 by default. 

More details about this parameter are here: https://typesense.org/docs/0.18.0/api/#search-collection (you want to scroll down to the table that describes all the parameters).

So if you need to query more than 250 rows in a multi-field search, you need to adjust the `max_values` parameter accordingly. If you don't mind disabling the performance optimization, you can also set `max_hits=all` which will restore the previous behavior. 

That said, a couple of people have run into this gotcha when upgrading. So we are now working on a fix to "auto-expand" the max_hits parameter as more pages are requested. Until then, setting `max_hits` is the solution.

**Comment by v-byte-cpu** [comment]
@jasonbosco documentation for `max_hits` states that default value for `max_hits` is 500. 

> (shouldn't be an issue for datasets lesser than say 5M records)

Maybe then set `max_hits` limit to this value ?

**Comment by jasonbosco** [comment]
@v-byte-cpu Yes, I meant `max_hits`. 

> Maybe then set max_hits limit to this value (instead of 250) or max_hits=all by default ?

Fair point. Good news is that we now have a way to actually do away with `max_hits` - one less config param to worry about. `v0.19.0` will be out in a few weeks with this update.

**Comment by jasonbosco** [comment]
Just published v0.19.0 which deprecated `max_hits`. So you can just remove that parameter from your search query going forward.
